core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -166,9 +167,16 @@ static MetaData updateMetaDataWithRoutingTable(MetaData oldMetaData, RoutingTabl
                 if (oldPrimary == null) {
                     throw new IllegalStateException("missing primary shard for " + newShardTable.shardId());
                 }
-                if (newPrimary.assignedToNode() &&
-                        newPrimary.isSameAllocation(oldPrimary) == false &&
-                        (oldPrimary.relocating() && newPrimary.isSameAllocation(oldPrimary.buildTargetRelocatingShard())) == false) {
+                // we update the primary term on initial assignment or when a replica is promoted. Most notably we do *not*
+                // update them when a primary relocates
+                if (newPrimary.unassigned() ||
+                        newPrimary.isSameAllocation(oldPrimary) ||
+                        // we do not use newPrimary.isTargetRelocationOf(oldPrimary) because that one enforces newPrimary to
+                        // be initializing. However, when the target shard is activated, we still want the primary term to staty
+                        // the same
+                        (oldPrimary.relocating() && newPrimary.isSameAllocation(oldPrimary.buildTargetRelocatingShard()))) {
```

Comment:
I believe we can safely expand the semantics of `isRelocationTargetOf` to also incorporate this case. This means changing `isRelocationTarget`, `isRelocationTargetOf` and `isRelocationSourceOf` such that it not only accounts for `INITIALIZING` state. (It is a safe change as a STARTED shard can never be a relocation source and RELOCATING shard can never be a relocation target).

For example, replacing

```
public boolean isRelocationTargetOf(ShardRouting other) {
    return this.allocationId != null && other.allocationId != null && this.state == ShardRoutingState.INITIALIZING &&
        this.allocationId.getId().equals(other.allocationId.getRelocationId());
```

by

```
public boolean isRelocationTargetOf(ShardRouting other) {
    return this.allocationId != null && other.allocationId != null && other.state == ShardRoutingState.RELOCATING  &&
        this.allocationId.getId().equals(other.allocationId.getRelocationId());
```

Can be a follow-up, just something to think about.